server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -485,13 +486,10 @@ private void recoverFromTranslogInternal(TranslogRecoveryRunner translogRecovery
         // note: if opsRecovered == 0 and we have older translogs it means they are corrupted or 0 length.
         assert pendingTranslogRecovery.get() : "translogRecovery is not pending but should be";
         pendingTranslogRecovery.set(false); // we are good - now we can commit
-        if (opsRecovered > 0) {
-            logger.trace("flushing post recovery from translog: ops recovered [{}], current translog generation [{}]",
-                opsRecovered, translog.currentFileGeneration());
-            commitIndexWriter(indexWriter, translog);
-            refreshLastCommittedSegmentInfos();
-            refresh("translog_recovery");
-        }
+        logger.trace(() -> new ParameterizedMessage(
+                "flushing post recovery from translog: ops recovered [{}], current translog generation [{}]",
+                opsRecovered, translog.currentFileGeneration()));
+        flush(false, true);
```

Comment:
I am a bit concerned about the [locking order](https://github.com/elastic/elasticsearch/pull/57350/files#diff-0a8902003e3ef16c5ca38aa269765578R1635) in`flush` (readLock -> flushLock) and `recoverFromTranslog` (flushLock -> readLock). I think we can remove the `flushLock` in `recoverFromTranslog` as it's safe to flush during recovery.